### ISSUE TITLE
Undocumented breaking change between 0.10 and 0.12?

### ISSUE THREAD

**Issue by lthms** [mirleft/ocaml-tls#427]
Hi!

First, thank you for this library. You are probably aware It is being used by the [Tezos blockchain](https://gitlab.com/tezos/tezos). As of now, the version used in Tezos is `0.10.6`, and we are in the process of [updating it to `0.12.8`](https://gitlab.com/tezos/tezos/-/merge_requests/2687?commit_id=4dcf6304ecc07208dd77acc7d8e995d196a2fa26).

The result of our effort is that Tezos builds correctly with this new version of TLS, but the tests fail with these sort of errors:

```
Warning:
  Failed to acquire the protocol version from the node
  Rpc request failed:
     - meth: GET
     - uri: https://localhost:18730/chains/main/blocks/head/protocols
     - error: Unable to connect to the node: "TLS failure: authentication failure: leaf certificate X.509 certificate
version 3
serial 1
algorithm RSA-SHA256
issuer /CN=Easy-RSA CA
valid from 2018-03-30 13:07:24 +00:00 until 2028-03-27 13:07:24 +00:00
subject /CN=tezos
extensions subjectKeyIdentifier b4 c2 ab c3 f6 64 80 94  43 46 7f 40 25 e4 d1 cf
                                01 33 44 da
keyUsage key encipherment, digital signature subjectAlternativeName dns tezos
 basicConstraints CA false depth none
authorityKeyIdentifier identifier 5e 27 08 3b 81 1d fa 05  cc d3 94 d4 2b 9b 92 5b
                                  3b f9 ea a1
issuer directory /CN=Easy-RSA CA  serial 15368070346713906475 
extendedKeyUsage server authentication
 does not contain the name localhost"
```

Unfortunately, I am no expert in TLS in general, but I am wondering: since the code builds, does it mean `tls` has become a bit more strict wrt. certificate validation? The changelog does not mention this, but maybe I have missed something? The certificate used in the test-suit can be found [here](https://gitlab.com/tezos/tezos/-/blob/master/tests_python/tools/constants.py#L96)

Thank you by advance for any help you could provide me on this, I have tried to get my head around these errors, but as of now, without success.

**Comment by hannesm** [comment]
Thanks for your report. I don't know much about "tezos" (and what kind of authentication they use), but indeed this is a certificate validation error. To shorten the message 'error: Unable to connect to the node: "TLS failure: authentication failure: leaf certificate [...] does not contain the name localhost"'.

This is pretty informative, even the X.509 certificate data is dumped, notably with: "CN=tezos" and "subjectAlternativeName dns tezos".

TL;DR: I have no clue why it worked before (and how) -- authentication is handled by the x509 library (an [X509.Authenticator.t](https://mirleft.github.io/ocaml-x509/doc/x509/X509/Authenticator/index.html#type-t) is provided to ocaml-tls (via Conig.client)). It looks like you try to connect to a host named `localhost` that presents a certificate that is issued to `tezos`, and the validation fails.

For testing, it may be sensible to use an authenticator that accepts everything (`let authenticator ~host:_ _certs = Ok None in Tls.Config.client ~authenticator ()` -- this won't do any name validation). For deployment this depends on the security model and how keys & certificates are managed.

Maybe "BREAKING Authenticator.null was removed" (from https://github.com/mirleft/ocaml-x509/releases/tag/v0.10.0) is the Changelog entry you're looking for (see https://github.com/mirleft/ocaml-x509/issues/130 for rationale).

**Comment by anmonteiro** [comment]
@lthms to mirror the previous behavior you can also just write a "null authenticator" like the following:

```ocaml
let null_auth ~host:_ _ = Ok None
```

**Comment by avsm** [comment]
It's probably worth a transitive CHANGES entry for this in `tls` as well -- this seems to be a common FAQ for people upgrading ocaml-tls.

**Comment by hannesm** [comment]
First of all, I understand the pain of upgrading packages (which "randomly" change their API etc.).

@lthms 
> I don’t know if this change in `x509` could justify a changelog note for `tls` as well.

I'm not convinced, since there's no tight coupling between x509 and tls (meaning a release of tls does not require a specific version of x509 - it is not clear where to put the Changelog entry) -- even further: a library that uses `tls` has used `x509` to a certain degree as well (since config requires them to do so).

IMHO instead we should investigate in tooling that composes a single Changes from a number of opam packages (i.e. odig with a diff mode on two switches) -- but this is not in scope for either this issue nor for ocaml-tls.

@avsm 
> this seems to be a common FAQ for people upgrading ocaml-tls.

I don't know, but this was the first of such issues AFAICT. Maybe I'm missing issues (or some other communication channels where people ask such questions)?

@anmonteiro 
> `let null_auth ~host:_ _ = Ok None`

I suggested this as well in https://github.com/mirleft/ocaml-tls/issues/427#issuecomment-811245882 - and want to add again the warning that this leads to communication setups where a [person-in-the-middle attack](https://en.wikipedia.org/wiki/Person-in-the-middle_attack) is achievable --> not recommended for production code.


I'll close this issue, in case there's more to discuss, please comment, reopen or discuss on discuss.ocaml.org.
